lib/SPIRV/SPIRVUtil.cpp

```diff
@@ -1639,6 +1639,281 @@ bool checkTypeForSPIRVExtendedInstLowering(IntrinsicInst *II, SPIRVModule *BM) {
+    for (Instruction *I : ToErase) {
+      assert(I->hasNUses(0));
+      I->eraseFromParent();
+    }
+  };
+  // Order of erasing is important.
+  Erase(Loads);
+  Erase(GEPs);
+  Erase(Casts);
+
+  return true;
+}
+
+bool lowerBuiltinVariablesToCalls(Module *M) {
+  std::vector<GlobalVariable *> WorkList;
+  for (auto I = M->global_begin(), E = M->global_end(); I != E; ++I) {
+    SPIRVBuiltinVariableKind Kind;
+    if (!isSPIRVBuiltinVariable(&(*I), &Kind))
+      continue;
+    if (!lowerBuiltinVariableToCall(&(*I), Kind))
+      return false;
+    WorkList.push_back(&(*I));
+  }
+  for (auto &I : WorkList) {
+    I->eraseFromParent();
+  }
+
+  return true;
+}
+
+/// Translates OCL/SPR-IR work-item builtin functions to SPIRV builtin
+/// variables.
+/// Function like get_global_id(i) -> x = load GlobalInvocationId; extract x, i
+/// Function like get_work_dim() -> load WorkDim
+void transWorkItemBuiltinsToVariables(Module *M) {
+  LLVM_DEBUG(dbgs() << "Enter transWorkItemBuiltinsToVariables\n");
+  std::vector<Function *> WorkList;
+  for (auto &I : *M) {
+    StringRef DemangledName;
+    if (!oclIsBuiltin(I.getName(), DemangledName))
+      continue;
+    LLVM_DEBUG(dbgs() << "Function demangled name: " << DemangledName << '\n');
+    std::string BuiltinVarName;
+    SPIRVBuiltinVariableKind BVKind;
+    // OCL builtin -> variable
+    if (!SPIRSPIRVBuiltinVariableMap::find(DemangledName.str(), &BVKind)) {
+      // SPV-IR builtin call -> variable
+      SmallVector<StringRef, 2> Postfix;
+      StringRef Name = dePrefixSPIRVName(DemangledName, Postfix);
+      if (!SPIRVBuiltInNameMap::rfind(Name.str(), &BVKind))
+        continue;
+    }
+    BuiltinVarName =
+        std::string(kSPIRVName::Prefix) + SPIRVBuiltInNameMap::map(BVKind);
+    LLVM_DEBUG(dbgs() << "builtin variable name: " << BuiltinVarName << '\n');
+    bool IsVec = I.getFunctionType()->getNumParams() > 0;
+    Type *GVType =
+        IsVec ? FixedVectorType::get(I.getReturnType(), 3) : I.getReturnType();
+    auto *BV = new GlobalVariable(
+        *M, GVType, true, GlobalValue::ExternalLinkage, nullptr, BuiltinVarName,
+        0, GlobalVariable::NotThreadLocal, SPIRAS_Input);
+    std::vector<Instruction *> InstList;
+    for (auto UI = I.user_begin(), UE = I.user_end(); UI != UE; ++UI) {
+      auto *CI = dyn_cast<CallInst>(*UI);
+      assert(CI && "invalid instruction");
+      const DebugLoc &DLoc = CI->getDebugLoc();
+      Instruction *NewValue = new LoadInst(GVType, BV, "", CI);
+      if (DLoc)
+        NewValue->setDebugLoc(DLoc);
+      LLVM_DEBUG(dbgs() << "Transform: " << *CI << " => " << *NewValue << '\n');
+      if (IsVec) {
+        NewValue =
+            ExtractElementInst::Create(NewValue, CI->getArgOperand(0), "", CI);
+        if (DLoc)
+          NewValue->setDebugLoc(DLoc);
+        LLVM_DEBUG(dbgs() << *NewValue << '\n');
+      }
+      NewValue->takeName(CI);
+      CI->replaceAllUsesWith(NewValue);
+      InstList.push_back(CI);
```

Comment:
I guess that if you switch `WorkList` to store `Value *`, then you will be able to re-use it here and reduce the code a little bit by performing a cleanup in a single place